Support arbitrary node attributes on paragraphs

stadelmanma opened the pull request:
<s>This lays the ground work for support of the style attribute by allowing any attributes set on an HTML node to be passed onto the WordML paragraph node. It also removes the need of the `ListParagraph` node subclass since we can now just pass the `numPr` attributes directly through.

Currently all attributes get shoved into the following template `<w:%s w:val="%s" />`, this will likely change when I implement styles because of tags like `</w:b>`. Nested attributes can be achieved by storing the value of your attribute in square brackets, i.e. `"[ilvl: #{@builder.ilvl}; numId: #{@definition.numid};]"`. 

I didn't intend for this to be advertised as a public API since it is primarily meant to serve our own internal needs but I didn't prevent it by white listing accepted attributes (or clearing them and setting our own). I figure there might be some good use cases out there that i don't know about.
</s>
Implementing the `style=` tag will make a pretty large diff in of itself just from the required boiler plate and creation of tests so I decided to split the PRs up again. 

(This PR conflicts with #49 in two files so which ever one you merge first, I'll fix the other)

stadelmanma commented:
@senny I have changed the previous logic using the string pattern match for nesting attributes to use nested Arrays/Hashes instead. While the other approach seemed pretty slick since it worked with the native Nokogiri node attributes my revised method is much more robust, and logical. 

The previous syntax was: `'numPr' => "[ilvl: #{@builder.ilvl}; numId: #{@definition.numid};]"`
And has been replaced by: 

```ruby
{
  'numPr' => [
    { 'ilvl' => @builder.ilvl }, { 'numId' => @definition.numid }
  ]
}
```

stadelmanma commented:
To clarify things, since I forgot to strike out the first half of that comment as well. No node attributes passed onto the WordXML node anymore except the `pStyle` attribute which we set based on the HTML tag. In the case of lists I pass the `numPr` props shown above.

senny commented on lib/sablon/html/ast.rb:
```diff
@@ -8,6 +8,26 @@ def accept(visitor)
+      private
+
+      # processes attributes defined on the node into wordML property syntax
+      def process_properties
```
Can we push this logic into a `ParagraphProperties` class? I'm thinking something along the lines of:

```
ParagraphProperties.new(
  'numPr' => [
    { 'ilvl' => 1 }, { 'numId' => 24 }
  ]
)
```

When constructing a Paragraph node one would instantiate it directly with a `ParagraphProperties` instance:

```ruby
properties = ParagraphProperties.new({ ... })
Paragraph.new(properties, runs)
```

The `ParagraphProperties` class would have some convenience methods to interact with the underlying data:

```ruby
props = ParagraphProperties.new({ ... })
props.style # => 'Normal'
props.to_docx # => <w:.....
```

senny replied:
It would also be nice to have a specific test-case exercising the ParagraphProperties class.

senny commented on lib/sablon/html/ast.rb:
```diff
@@ -45,23 +65,14 @@ def inspect
-
       def to_docx
-        PATTERN % [style, ppr_docx, runs.to_docx]
+        "<w:p>#{ppr_docx}#{runs.to_docx}</w:p>"
```
Following the thought of a `ParagraphProperties` object, this becomes:

```ruby
"<w:p>#{properties.to_docx}#{runs.to_docx}</w:p>"
```

And we can get rid of `ppr_docx`.

stadelmanma commented:
@senny I have added the properties class but in favor of doing things in a more flexible manner I called it `NodeProperties` instead. When calling the `to_docx` method on the properties instance you pass in the base XML tag name (`w:p` for paragraphs) to get the proper markup for that specific node. 

I implemented a `[](key)` and `[]=(key,value)` method for convenient access to the underlying hash since that is the only data it will ever store.

There is now a robust test case in converter_test.rb that throughly checks the logic and the recursive nature of `transform_attr`.

senny commented on test/html/converter_test.rb:
```diff
@@ -328,6 +328,62 @@ def setup
     @converter.instance_variable_set(:@numbering, Sablon::Environment.new(nil).numbering)
   end
 
+  def test_node_properties_converison
```
Instead of this single huge method, Can we create a new TestCase class in this file and split each example separated by a comment into a method?

```
class NodePropertiesTest < Sablon::TestCase
  def test_empty_properties
  end

  def test_simple_properties
  end

  ...
end
```

senny commented on lib/sablon/html/ast.rb:
```diff
@@ -10,6 +10,54 @@ def self.node_name
+        @properties[key] = value
+      end
+
+      def to_docx(parent_tag)
```
I don't like the fact that this method has a different signature than other `to_docx` methods. All of this code passes `'w:p'`. Why not go with `ParagraphProperties` and if necessary extend from a more abstract `NodeProperties` class (only necessary once we see more specializations). Another approach would be to keep `NodeProperties` but pass this key when the object is instantiated. To simplify things for call-sites we can provide factory methods like `.paragraph`.

```ruby
NodeProperties.paragraph({})
```

stadelmanma commented:
@senny, updated `to_docx` call signature to match. I changed the `NodePoperties.initialize` method to accept the required tagname. Instead of implicitly appending a 'Pr' to the parent tag I thought it might be better to be explicit with things and pass the tagname in as a whole (i.e 'pPr' instead)

I think subclassing an abstract NodeProperties would be overkill to just change the tagname used.

stadelmanma commented:
Requested changes added. The only place `w:pPr` was getting used outside of tests was inside the Paragraph class itself. I added a test for the `NodeProperties.paragraph` method but I didn't convert all of the other unit tests over to use it. I didn't see a need to route all of our unit tests through the factory when the underlying code is 'tagname' agnostic and we want to maintain that behavoir for future expansion.

senny commented:
thank you.
